sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java

```diff
@@ -3,24 +3,56 @@
 
 package com.azure.core.http.policy;
 
+import com.azure.core.http.HttpHeaders;
 import com.azure.core.http.HttpPipelineCallContext;
 import com.azure.core.http.HttpPipelineNextPolicy;
 import com.azure.core.http.HttpResponse;
 import reactor.core.publisher.Mono;
 
+import java.util.Objects;
 import java.util.UUID;
 
 /**
- * The pipeline policy that puts a UUID in the request header. Azure uses the request id as
+ * The pipeline policy that puts a UUID or user provided id in the request header. Azure uses the request id as
  * the unique identifier for the request.
+ * @see RequestIdPolicyOptions
  */
 public class RequestIdPolicy implements HttpPipelinePolicy {
     private static final String REQUEST_ID_HEADER = "x-ms-client-request-id";
 
+    private final RequestIdPolicyOptions requestIdPolicyOptions;
+
+    /**
+     * Creates {@link RequestIdPolicy} with default {@link ExponentialBackoff} as {@link RetryStrategy}and use
+     * 'retry-after-ms' in {@link HttpResponse} header for calculating retry delay.
+     */
+    public RequestIdPolicy() {
+        this.requestIdPolicyOptions = null;
+    }
+
+    /**
+     * Creates a {@link RequestIdPolicy} with the provided {@link RequestIdPolicyOptions}.
+     * User can specify {@link java.util.function.Function} to dynamically generated id for various id headers.
+     *
+     * @param requestIdPolicyOptions with given {@link RetryPolicyOptions}.
+     * @throws NullPointerException if {@code RequestIdPolicyOptions} is {@code null}.
+     */
+    public RequestIdPolicy(RequestIdPolicyOptions requestIdPolicyOptions) {
+        this.requestIdPolicyOptions = Objects.requireNonNull(requestIdPolicyOptions,
+            "'retryPolicyOptions' cannot be null.");
+    }
+
+
     @Override
     public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
-        String requestId = context.getHttpRequest().getHeaders().getValue(REQUEST_ID_HEADER);
-        if (requestId == null) {
+        if (!Objects.isNull(requestIdPolicyOptions)) {
+            HttpHeaders httpHeaders =  requestIdPolicyOptions.getIdHeaderSupplier().get();
+            if (!Objects.isNull(httpHeaders)) {
+                httpHeaders.stream().forEach(httpHeader ->
+                    context.getHttpRequest().getHeaders().put(httpHeader.getName(), httpHeader.getValue())
```

Comment:
This should check for existing headers with the same name in the context. If it already exists, it should not be overwritten.

Reply:
I think this is a good place to have a discussion, can take it offline if needed.

@JonathanGiles is there any guidelines around whether a pipeline policy should overwrite or only add if not exists?

Reply:
I am also unsure about guideline on this. I did hear that every new request  ( including retry ) to Azure service should use new new Request ID.  But at the same time it is "optional" in Azure Docs
https://docs.microsoft.com/en-us/rest/api/azure/devops/symbol/requests/create%20requests?view=azure-devops-rest-5.1

Reply:
I just spoke to Jeffery on this  and his comments are "Every retry should have the same ID. This way, we can tell how many retries occurred for a single logical operation."